rbd/group_snap_info.go

```diff
@@ -8,17 +8,83 @@ package rbd
 #include <stdlib.h>
 #include <rbd/librbd.h>
 
+// Types and constants are copied from librbd.h with added "_" as prefix. This
+// prevents redefinition of the types on librbd versions that have them
+// already.
+
+typedef enum {
+  _RBD_GROUP_SNAP_NAMESPACE_TYPE_USER = 0
+} _rbd_group_snap_namespace_type_t;
+
+typedef struct {
+  char *image_name;
+  int64_t pool_id;
+  uint64_t snap_id;
+} _rbd_group_image_snap_info_t;
+
+typedef struct {
+  char *id;
+  char *name;
+  char *image_snap_name;
+  rbd_group_snap_state_t state;
+  _rbd_group_snap_namespace_type_t namespace_type;
+  size_t image_snaps_count;
+  _rbd_group_image_snap_info_t *image_snaps;
+} _rbd_group_snap_info2_t;
+
+// rbd_group_snap_get_info_fn matches the rbd_group_snap_get_info function signature.
+typedef int(*rbd_group_snap_get_info_fn)(rados_ioctx_t group_p,
+                                         const char *group_name,
+                                         const char *snap_name,
+                                         _rbd_group_snap_info2_t *snaps);
+
+// rbd_group_snap_get_info_dlsym take *fn as rbd_group_snap_get_info_fn and
+// calls the dynamically loaded rbd_group_snap_get_info function passed as 1st
+// argument.
+static inline int rbd_group_snap_get_info_dlsym(void *fn,
+                                                rados_ioctx_t group_p,
+                                                const char *group_name,
+                                                const char *snap_name,
+                                                _rbd_group_snap_info2_t *snaps) {
+  // cast function pointer fn to rbd_group_snap_get_info and call the function
+  return ((rbd_group_snap_get_info_fn) fn)(group_p, group_name, snap_name, snaps);
+}
+
+// rbd_group_snap_get_info_cleanup_fn matches the rbd_group_snap_get_info_cleanup function signature.
+typedef int(*rbd_group_snap_get_info_cleanup_fn)(_rbd_group_snap_info2_t *snaps);
+
+// rbd_group_snap_get_info_cleanup_dlsym take *fn as rbd_group_snap_get_info_cleanup_fn and
+// calls the dynamically loaded rbd_group_snap_get_info_cleanup function passed as 1st
+// argument.
+static inline int rbd_group_snap_get_info_cleanup_dlsym(void *fn,
+                                                _rbd_group_snap_info2_t *snaps) {
+  // cast function pointer fn to rbd_group_snap_get_info_cleanup and call the function
+  return ((rbd_group_snap_get_info_cleanup_fn) fn)(snaps);
+}
```

Comment:
I wish we could avoid some of that boilerplate... But I don't have a suggestion right now, how to improve it. So I guess we just keep it like that for now, which probably means forever. 😅